### PR TITLE
Initial contents for BinaryFormatter security guide

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
## Summary

This represents the initial contents for the `BinaryFormatter` security guide, which will be linked to from https://aka.ms/binaryformatter and is part of the long-term `BinaryFormatter` obsoletion strategy.

This is intended to be a __living document__ and will eventually grow to include very detailed guidance for specific scenarios, including code samples. By "living document" I mean that we'll probably update it on a monthly cadence until we're satisfied that the guidance covers the necessary use cases. (See https://github.com/dotnet/docs/pull/19442#issuecomment-657141592 for a partial list of topics.)

If this repo isn't the proper place for such a document please let me know and I can figure out where else to put it. But given all of the other serialization-related content in this repo, it seemed like a natural fit.

I'm pushing the contents here now so that we can iterate on the formatting or wording if needed. There is already signoff on the high-level concepts covered by the document.

## Todo before checkin

- [ ] Update the table at the top, including with keywords and a real asset id
- [ ] Add links from `BinaryFormatter` API page and related docs to this
- [ ] Include code samples for using `XmlSerializer` and `JsonSerializer`
- [ ] Fix wording in other docs to clarify that `BinaryFormatter` cannot be made safe, even with a custom binder (see https://github.com/dotnet/docs/pull/19442#issuecomment-657141781 for initial list)

/cc @Rick-Anderson @gewarren

### CODE REVIEW

**Comment by GrabYourPitchforks** [comment]
Topics we'll want to cover, with samples:

- [ ] Serializing `Exception` graphs
- [ ] Polymorphic deserialization with _System.Text.Json_
- [ ] Reasoning about cycles, data complexity, algorithmic attacks
- [ ] Deserializing dictionaries safely
- [ ] Dangerous `Type.GetType` calls

**Comment by GrabYourPitchforks** [comment]
Docs we need to update because they give improper security guidance:

- [ ] https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.formatters.binary.binaryformatter - to link to this document
- [ ] https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.formatters.binary.binaryformatter.binder - to clarify that a SerializationBinder is not a long-term security solution
- [ ] https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.serializationbinder - to remove the "SerializationBinder can be used for security" line (w.r.t. BinaryFormatter / NetDataContractSerializer)
- [ ] https://docs.microsoft.com/en-us/dotnet/framework/wcf/feature-details/controlling-serialization-and-deserialization-with-serializationbinder - to clarify that this shouldn’t be used for security
- [ ] https://docs.microsoft.com/en-us/previous-versions/dotnet/netframework-4.0/5dxse167(v=vs.100) – clarifying that a TypeFilterLevel shouldn’t be used for security
- [ ] https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.formatters.typefilterlevel - clarifying that a TypeFilterLevel shouldn’t be used for security
- [ ] https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.netdatacontractserializer
- [ ] https://docs.microsoft.com/en-us/dotnet/api/system.web.ui.losformatter
- [ ] https://docs.microsoft.com/en-us/dotnet/api/system.web.ui.objectstateformatter
- [ ] https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.formatters.soap.soapformatter - to clarify that these are all dangerous serializers and should link to this doc
- [ ] https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2300
- [ ] https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2301
- [ ] https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2302 - to clarify that SerializationBinder is not recommended as a long-term mitigation strategy
- [ ] https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2310
- [ ] https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2311
- [ ] https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2312 - same as above, but for NetDataContractSerializer
- [ ] https://docs.microsoft.com/en-us/dotnet/framework/wcf/feature-details/security-considerations-for-data – remove text "It is possible to make it secure by writing a secure, type-limiting type binder"

File: docs/standard/serialization/binaryformatter-security-guide.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+This document applies to the following framework versions:
+
+* .NET Framework (all versions)
+* .NET Core 2.1 - 3.1
+* .NET 5.0+
+
+## Background
+
+> __Summary__: The `BinaryFormatter` type is dangerous and is not recommended for data processing. Applications should stop using it at their earliest opportunity, even if they believe the data they're processing to be trustworthy.
```
Comment (Youssef1313) [inline docs/standard/serialization/binaryformatter-security-guide.md:19]:
```suggestion
> __Summary__: The <xref:System.Runtime.Serialization.Formatters.Binary.BinaryFormatter> type is dangerous and is not recommended for data processing. Applications should stop using it at their earliest opportunity, even if they believe the data they're processing to be trustworthy.
```

File: docs/standard/serialization/binaryformatter-security-guide.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+> This document also applies to the `SoapFormatter`, `NetDataContractSerializer`, `LosFormatter`, and `ObjectStateFormatter` types.
```
Comment (Youssef1313) [inline docs/standard/serialization/binaryformatter-security-guide.md:21]:
```suggestion
> This document also applies to the <xref:System.Runtime.Serialization.Formatters.Soap.SoapFormatter>, <xref:System.Runtime.Serialization.NetDataContractSerializer>, <xref:System.Web.UI.LosFormatter>, and <xref:System.Web.UI.ObjectStateFormatter> types.
```

File: docs/standard/serialization/binaryformatter-security-guide.md (hunk 1 above)
Comment (Youssef1313) [reply docs/standard/serialization/binaryformatter-security-guide.md:19]: > Applications should stop using it at their earliest opportunity, even if they believe the data they're processing to be trustworthy.

As a normal reader, this is confusing for me, why would a trusted source be a problem?
If this is meant that I *may* think it's trusted while it's not, I'd indicate that clearly.

**Comment by Youssef1313** [comment]
There are some roslyn's analyzers related to this, for example [CA2300](https://docs.microsoft.com/visualstudio/code-quality/ca2300) (and others too). You may like to reference them.

File: docs/standard/serialization/binaryformatter-security-guide.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+`BinaryFormatter.Deserialize` may be vulnerable to other attack categories, such as information disclosure or remote code execution. Utilizing features such as a custom [serialization binder](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.serializationbinder) may be insufficient to properly mitigate these risks. The possibility exists that a novel vulnerability will be discovered for which .NET cannot practically publish a security update. Consumers should assess their individual scenarios and consider their potential exposure to these risks.
```
Comment (Youssef1313) [inline docs/standard/serialization/binaryformatter-security-guide.md:35]:
```suggestion
`BinaryFormatter.Deserialize` may be vulnerable to other attack categories, such as information disclosure or remote code execution. Utilizing features such as a custom <xref:System.Runtime.Serialization.SerializationBinder> may be insufficient to properly mitigate these risks. The possibility exists that a novel vulnerability will be discovered for which .NET cannot practically publish a security update. Consumers should assess their individual scenarios and consider their potential exposure to these risks.
```

File: docs/standard/serialization/binaryformatter-security-guide.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+We recommend that `BinaryFormatter` consumers perform individual risk assessments on their applications. It is the consumer's sole responsibility to determine for itself whether to utilize this API. Consumers should assess any security, technical, reputational, and legal risks (including regulatory requirements) that may accompany using `BinaryFormatter` within their applications.
+
+## Preferred alternatives
+
+.NET offers several in-box serializers which can handle untrusted data safely.
+
+The in-box types [`XmlSerializer`](https://docs.microsoft.com/en-us/dotnet/api/system.xml.serialization.xmlserializer) and [`DataContractSerializer`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.datacontractserializer) (not to be confused with `NetDataContractSerializer`) can be used to serialize object graphs into and from XML safely.
```
Comment (Youssef1313) [inline docs/standard/serialization/binaryformatter-security-guide.md:43]:
```suggestion
The in-box types <xref:System.Xml.Serialization.XmlSerializer) and <xref:System.Runtime.Serialization.DataContractSerializer> (not to be confused with `NetDataContractSerializer`) can be used to serialize object graphs into and from XML safely.
```

File: docs/standard/serialization/binaryformatter-security-guide.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+The [_System.Text.Json_](https://docs.microsoft.com/en-us/dotnet/api/system.text.json) APIs can be used to serialize object graphs into JSON safely.
```
Comment (Youssef1313) [inline docs/standard/serialization/binaryformatter-security-guide.md:45]:
```suggestion
The <xref:System.Text.Json> APIs can be used to serialize object graphs into JSON safely.
```

**Comment by HurricanKai** [comment]
In my opinion `System.Buffers.Binary.BinaryPrimitives` should be considered as an alternative next to XML & JSON for those that require binary data.
Especially considering the example of a data file, JSON/XML may not be a viable alternative, as JSON/XML documents can grow very large, and JSON/XML would simply bloat the file too much, also inflating potential load times.
BinaryPrimitives can only encode primitive types, which makes it a more complicated solution, but in some cases, this may be worth it.

**Comment by GrabYourPitchforks** [comment]
@HurricanKai good suggestion! I'd probably recommend `BinaryReader` and `BinaryWriter` instead of `BinaryPrimitives` since they're more friendly to streaming scenarios, but it still matches the overall spirit of your suggestion.

**Comment by Rick-Anderson** [comment]
Moved to #19445

**Pull request closed without merging**